### PR TITLE
Fully support base type mapping names with precision/scale/size specified separately

### LINKED ISSUE #15159: Fluent mapping with MaxLength giving type with no length
I'm trying to map a database from database first to code first. While I'm using `IEntityTypeConfiguration` classes to specify my mappings one problem occurs, the mapping of a char field to the database with a maxlength gives a RawSql of char with no specification of length.

Same for NCHAR and VARBINARY 

### Steps to reproduce

```c#
builder.Property(e => e.NotificationsLot).HasColumnType("char").HasMaxLength(5).IsFixedLength();
```

which compiles to

```c#
NotificationsLot = table.Column<string>(type: "char", fixedLength: true, maxLength: 5, nullable: true),
```

which in turn turns into

```sql
[NotificationsLot] char NULL,
```

### Further technical details
EF Core version: 2.2.3
Database Provider: Microsoft.EntityFrameworkCore.SqlServer
Operating system: Windows 10 Pro
IDE: Visual Studio 2019 Preview 16.0.0 Preview 4.3

### LINKED ISSUE #21061: HasPrecision() does not override [Column(TypeName=data_type(precision, scale))]
If you set up the code below:

```C#
    public class PrecisionContext01 : DbContext
    {
        public DbSet<Precision> Precisions { get; set; }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            optionsBuilder.UseSqlServer(@"your_connection_string");
        }

        protected override void OnModelCreating(ModelBuilder builder)
        {
            builder.Entity<Precision>()
                .Property(p => p.Score).HasPrecision(14, 2);
        }
    }

    public class Precision
    {
        public int Id { get; set; }
        [Column(TypeName = "decimal(17,5)")]
        public decimal Score { get; set; }
    }
```

And then do `Add-Migration` followed by `Script-Migration` you get the following SQL. I.e. the column type obeyed the annotation and not the fluent API.

```SQL
CREATE TABLE [Precisions] (
    [Id] int NOT NULL IDENTITY,
    [Score] decimal(17,5) NOT NULL,
    CONSTRAINT [PK_Precisions] PRIMARY KEY ([Id])
);```

### LINKED ISSUE #21140: Not sure if it is a bug or I'm doing something wrong: Scaffolding doesn't work for "varchar" data type
In my service, trying to use EF Core DBContext for service to access to database. In my Web API controller, I use constructor injection to get the db context, and when debugging, I can see the dbContext was injected into the controller and can be used. But the `entity` in the DbContext always has exception:
 `System.ArgumentException: Data type 'VARCHAR' is not supported in this form. Either specify the length explicitly in the type name, for example as 'VARCHAR(16)', or remove the data type and use APIs such as HasMaxLength to allow EF choose the data type.`

I did the following config on DbContext for my asp.net core web api in service fabric.

MyWebService.cs, config the dbcontext, and inject it into service. piece of code is like:
``` C#
.ConfigureServices(services => services.AddDbContext<MyDbContext>(options =>
   {
      options.UseSqlServer("<sql_db_connectionString>", sqlServerOptionsAction: sqlOptions =>
      {
         sqlOptions.EnableRetryOnFailure(
             maxRetryCount: 5,
             maxRetryDelay: 10,
             errorNumbersToAdd: null
         );
      });
   }))
```
MyDbContext.cs
``` C#
public class MyDbContext : DbContext
{
    public MyDbContext(DbContextOptions<MyDbContext> options) : base(options)
    {}
    public virtual DbSet<Features> Features { get; set; }
}
```
The entity `Features` class, which is mapping to a table in database
``` C#
[Table("Features")]
public class Features
{
    [Column(TypeName = "VARCHAR")]
    [MaxLength(50)]
    public string FeatureName { get; set; }
    // other properties...
}
```
And in Web API controller, use constructor injection to get the dbcontext:
``` C#
public class MyController : ControllerBase
{
    private MyDbContext myDbContext;
    public MyController(MyDbContext dbContext)
    {
        myDbContext = dbContext;
    }

    public string GetData()
    {
      var result = from strings in this.myDbContext.Features .........
    }
}
```

The above config way( way1 )--- using attribute in Entity gives me the VARCHAR not supported data type exception.

I also tried to use fluent API (way 2) to config entity property, for which I remove the attribute for the entity property like the following, but still same exception.
``` C#
public class MyDbContext : DbContext
{
    public MyDbContext(DbContextOptions<MyDbContext> options) : base(options)
    {}
    public virtual DbSet<Features> Features { get; set; }

    protected override void OnModelCreating(ModelBuilder modelBuilder)
    {
        modelBuilder.Entity<Features>()
            .Property(e => e.FeatureName)
            .HasColumnName("FeatureName")
            .HasColumnType("varchar(50)");
    }
}

[Table("Features")]
public class Features
{
    public string FeatureName { get; set; }
    // other properties...
}

```
image

### Further technical details

EF Core version: 3.1.4
Database provider:  Microsoft.EntityFrameworkCore.SqlServer
Target framework: .netcoreapp 3.1
Operating system:
IDE: Visual Studio 2019 16.6.0

### PR DESCRIPTION
**Pull request by ajcvickers**
Fixes #21061
Fixes #21140
Fixes #15159

### CODE REVIEW

File: src/EFCore.Relational/Extensions/RelationalPropertyExtensions.cs [hunk 1]
```diff
@@ -254,7 +251,7 @@ public static string GetColumnType([NotNull] this IProperty property, StoreObjec
             var annotation = property.FindAnnotation(RelationalAnnotationNames.ColumnType);
             if (annotation != null)
             {
-                return (string)annotation.Value;
+                return property.FindRelationalTypeMapping()?.StoreType;
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Extensions/RelationalPropertyExtensions.cs:254]:
```suggestion
                return property.FindRelationalTypeMapping(storeObject)?.StoreType ?? (string)annotation.Value;
```

File: src/EFCore.SqlServer/Storage/Internal/SqlServerTypeMappingSource.cs [hunk 2]
```diff
@@ -65,7 +65,7 @@ public class SqlServerTypeMappingSource : RelationalTypeMappingSource
             = new SqlServerStringTypeMapping(unicode: true, fixedLength: true);
 
         private readonly SqlServerStringTypeMapping _textUnicodeString
-            = new SqlServerStringTypeMapping("ntext", unicode: true, sqlDbType: SqlDbType.NText);
+            = new SqlServerStringTypeMapping("ntext", unicode: true, sqlDbType: SqlDbType.NText, storeTypePostfix: StoreTypePostfix.None);
```
Comment (AndriySvyryd) [inline src/EFCore.SqlServer/Storage/Internal/SqlServerTypeMappingSource.cs:68]: The last two parameter names are redundant, same below

File: src/EFCore.SqlServer/Storage/Internal/SqlServerDateTimeTypeMapping.cs [hunk 3]
```diff
@@ -43,8 +44,16 @@ public class SqlServerDateTimeTypeMapping : DateTimeTypeMapping
         /// </summary>
         public SqlServerDateTimeTypeMapping(
             [NotNull] string storeType,
-            DbType? dbType = null)
-            : base(storeType, dbType)
+            DbType? dbType = null,
+            int? precision = null,
+            StoreTypePostfix storeTypePostfix = StoreTypePostfix.Precision)
```
Comment (AndriySvyryd) [inline src/EFCore.SqlServer/Storage/Internal/SqlServerDateTimeTypeMapping.cs:49]: When would this be called with a different `StoreTypePostfix`?

File: src/EFCore.SqlServer/Storage/Internal/SqlServerTypeMappingSource.cs (hunk 2 above)
Comment (ajcvickers) [reply src/EFCore.SqlServer/Storage/Internal/SqlServerTypeMappingSource.cs:68]: Note sure what you mean by this...

Do you want values specified for all parameters?

File: src/EFCore.SqlServer/Storage/Internal/SqlServerDateTimeTypeMapping.cs (hunk 3 above)
Comment (ajcvickers) [reply src/EFCore.SqlServer/Storage/Internal/SqlServerDateTimeTypeMapping.cs:49]: Removed

File: src/EFCore.SqlServer/Storage/Internal/SqlServerTypeMappingSource.cs (hunk 2 above)
Comment (AndriySvyryd) [reply src/EFCore.SqlServer/Storage/Internal/SqlServerTypeMappingSource.cs:68]:
```suggestion
            = new SqlServerStringTypeMapping("ntext", unicode: true, SqlDbType.NText, StoreTypePostfix.None);
```

File: src/EFCore.SqlServer/Storage/Internal/SqlServerDateTimeTypeMapping.cs (hunk 3 above)
Comment (AndriySvyryd) [reply src/EFCore.SqlServer/Storage/Internal/SqlServerDateTimeTypeMapping.cs:49]: Same for other mappings

File: src/EFCore.SqlServer/Storage/Internal/SqlServerTypeMappingSource.cs (hunk 2 above)
Comment (ajcvickers) [reply src/EFCore.SqlServer/Storage/Internal/SqlServerTypeMappingSource.cs:68]: Doesn't compile.

Comment (ajcvickers) [reply src/EFCore.SqlServer/Storage/Internal/SqlServerTypeMappingSource.cs:68]: size and fixedLength would need to be specified too.

**Pull request merged**